que.go

```diff
@@ -216,46 +216,56 @@ func (c *Client) LockJob(queue string) (*Job, error) {
 	}
 
 	j := Job{pool: c.pool, conn: conn}
-	err = conn.QueryRow("que_lock_job", queue).Scan(
-		&j.Queue,
-		&j.Priority,
-		&j.RunAt,
-		&j.ID,
-		&j.Type,
-		&j.Args,
-		&j.ErrorCount,
-	)
-	if err != nil {
-		c.pool.Release(conn)
-		if err == pgx.ErrNoRows {
-			return nil, nil
+
+	for {
+		err = conn.QueryRow("que_lock_job", queue).Scan(
+			&j.Queue,
+			&j.Priority,
+			&j.RunAt,
+			&j.ID,
+			&j.Type,
+			&j.Args,
+			&j.ErrorCount,
+		)
+		if err != nil {
+			c.pool.Release(conn)
+			if err == pgx.ErrNoRows {
+				return nil, nil
+			}
+			return nil, err
 		}
-		return nil, err
-	}
 
-	// Deal with race condition. Explanation from the Ruby Que gem:
-	//
-	// Edge case: It's possible for the lock_job query to have
-	// grabbed a job that's already been worked, if it took its MVCC
-	// snapshot while the job was processing, but didn't attempt the
-	// advisory lock until it was finished. Since we have the lock, a
-	// previous worker would have deleted it by now, so we just
-	// double check that it still exists before working it.
-	//
-	// Note that there is currently no spec for this behavior, since
-	// I'm not sure how to reliably commit a transaction that deletes
-	// the job in a separate thread between lock_job and check_job.
-	var ok bool
-	err = conn.QueryRow("que_check_job", j.Queue, j.Priority, j.RunAt, j.ID).Scan(&ok)
-	if err != nil {
-		c.pool.Release(conn)
-		if err == pgx.ErrNoRows { // encountered job race condition
-			return nil, nil
+		// Deal with race condition. Explanation from the Ruby Que gem:
+		//
+		// Edge case: It's possible for the lock_job query to have
+		// grabbed a job that's already been worked, if it took its MVCC
+		// snapshot while the job was processing, but didn't attempt the
+		// advisory lock until it was finished. Since we have the lock, a
+		// previous worker would have deleted it by now, so we just
+		// double check that it still exists before working it.
+		//
+		// Note that there is currently no spec for this behavior, since
+		// I'm not sure how to reliably commit a transaction that deletes
+		// the job in a separate thread between lock_job and check_job.
+		var ok bool
+		err = conn.QueryRow("que_check_job", j.Queue, j.Priority, j.RunAt, j.ID).Scan(&ok)
+		if err == nil {
+			return &j, nil
+		} else if err == pgx.ErrNoRows {
+			// Encountered job race condition; start over from the beginning.
+			// We're still holding the advisory lock, though, so we need to
+			// release it before resuming.  Otherwise we leak the lock,
+			// eventually causing the server to run out of locks.
+			//
+			// Also swallow the possible error, exactly like in Done.
+			var ok bool
```

Comment:
yeah, I don't have a great idea to improve this either. I'm pretty sure I modeled this after what the ruby version was doing. I mean even if you were to bubble up an error, the user couldn't necessarily do anything to fix it. And it's definitely an internal concern IMO that users shouldn't need to worry about unless it's an irrecoverable condition.

I wonder, if an error is encountered here, what's the likelihood that the conn would be thrown away by the underlying pgx driver?